lock-api/src/main/java/com/palantir/lock/client/LeasingTimelockClient.java

```diff
@@ -0,0 +1,154 @@
+        this.delegate = timelockService;
+        this.lockLeaseManager = lockLeaseManager;
+    }
+
+    public static LeasingTimelockClient create(TimelockServerInterface timelockServerInterface) {
+        return new LeasingTimelockClient(timelockServerInterface, LockLeaseManager.create());
+    }
+
+    @Override
+    public long getFreshTimestamp() {
+        return delegate.getFreshTimestamp();
+    }
+
+    @Override
+    public TimestampRange getFreshTimestamps(int numTimestampsRequested) {
+        return delegate.getFreshTimestamps(numTimestampsRequested);
+    }
+
+    @Override
+    public LockImmutableTimestampResponse lockImmutableTimestamp(IdentifiedTimeLockRequest request) {
+        return delegate.lockImmutableTimestamp(request);
+    }
+
+    @Override
+    public StartAtlasDbTransactionResponse startAtlasDbTransaction(IdentifiedTimeLockRequest request) {
+        return delegate.startAtlasDbTransaction(request);
+    }
+
+    @Override
+    public StartIdentifiedAtlasDbTransactionResponse startIdentifiedAtlasDbTransaction(
+            StartIdentifiedAtlasDbTransactionRequest request) {
+        long startTime = System.nanoTime();
+        LeasableStartIdentifiedAtlasDbTransactionResponse leasableResponse =
+                delegate.leasableStartIdentifiedAtlasDbTransaction(request);
+
+        StartIdentifiedAtlasDbTransactionResponse response = leasableResponse.getStartTransactionResponse();
+        Optional<Duration> leasePeriod = leasableResponse.getLeasePeriod();
+
+        updateLockLeases(response.immutableTimestamp().getLock(), startTime, leasePeriod);
+
+        return response;
+    }
+
+    @Override
+    public long getImmutableTimestamp() {
+        return delegate.getImmutableTimestamp();
+    }
+
+    @Override
+    public LockResponse lock(LockRequest request) {
+        long startTime = System.nanoTime();
+        LeasableLockResponse leasableResponse = delegate.leasableLock(request);
+
+        LockResponse lockResponse = leasableResponse.getLockResponse();
+        Optional<Duration> leasePeriod = leasableResponse.getLeasePeriod();
+
+        if (lockResponse.wasSuccessful()) {
+            updateLockLeases(lockResponse.getToken(), startTime, leasePeriod);
+        }
+
+        return lockResponse;
+    }
+
+    @Override
+    public WaitForLocksResponse waitForLocks(WaitForLocksRequest request) {
+        return delegate.waitForLocks(request);
+    }
+
+    @Override
+    public Set<LockToken> refreshLockLeases(Set<LockToken> tokens) {
+        Set<LockToken> validByLease = tokens.stream()
+                .filter(lockLeaseManager::isValid)
+                .collect(Collectors.toSet());
+
+        Set<LockToken> toRefresh = Sets.difference(tokens, validByLease);
+
+        long startTime = System.nanoTime();
+        LeasableRefreshLockResponse refreshLockResponse = delegate.leasableRefreshLockLeases(toRefresh);
```

Comment:
Some tokens that were valid due to leases may have expired by this point, given the potential length of the RPC, but that doesn't matter given our protocol (one could always have lost the locks immediately after having verified that they were held)